libraries/cbor/src/writer.rs

```diff
@@ -56,8 +56,14 @@ impl<'a> Writer<'a> {
                     }
                 }
             }
-            Value::Map(map) => {
-                self.start_item(5, map.len() as u64);
+            Value::Map(mut map) => {
+                map.sort_by(|a, b| a.0.cmp(&b.0));
+                let map_len = map.len();
+                map.dedup_by(|a, b| a.0.eq(&b.0));
+                if map_len != map.len() {
+                    return false;
+                }
```

Comment:
We are changing the previous behavior here. Before when using `cbor_map!` we would keep the last key-value pair for a given key. Now, it is unspecified. We could specify it using a stable sort and reversing the slice before deduping to preserve the previous behavior. But we probably don't care, so just mentioning. Another difference is that before we did not return false for duplicates. It was silently ignored. But there seems to be another discussion somewhere else about that. This is just to point out the 2 differences.

Reply:
My other response is [here](https://github.com/google/OpenSK/pull/303#discussion_r610458792). I think this issue is new because from a library perspective, the input format simply didn't allow duplicates, as it was a `BTreeMap`. Therefore, I wouldn't see that as changing previous behaviour, but making a choice for some new behaviour. if that makes sense?

Reply:
Yes it makes sense but I think it's only one part of the story. As I see it (only listing the relevant parts for this discussion):
- There is `cbor::write(value: cbor::Value, ...)` that is part of the public API. This function takes a `cbor::Value` which is an open data-type so changing from `BTreeMap` to `Vec` is an API change. As such an invariant that was enforced by the data-type is now removed. So we are free to specify the new behavior. That's your point.
- There is `cbor_map!{ k1 => v1, k2 => v2, ... }` that builds a `cbor::Value` out of a list of key-values which may contain duplicates and don't need to be sorted. The behavior was to silently use the last value for each duplicate key. It is also implicitly sorted due to the data-structure. My point is that this function changes behavior with the same API, so we are not really choosing a new behavior, but changing the existing one. I don't say we shouldn't change the behavior, I just say we should be aware that we are changing it and agree that's our choice. Ideally we should document the `cbor_map!` macro which currently doesn't have any documentation.

Reply:
The macros definitely need documentation. Added for maps and arrays. The literal types might also need some work later.

Reply:
Oh my bad, I forgot the fact that we reject duplicate keys now. So the difference in behavior in `cbor_map!` is not observable for non-error cases. Nice!